src/components/battery/BatteryController.cpp

```diff
@@ -97,3 +93,37 @@ void Battery::SaadcEventHandler(nrfx_saadc_evt_t const* p_event) {
 void Battery::Register(Pinetime::System::SystemTask* systemTask) {
   this->systemTask = systemTask;
 }
+
+uint8_t Battery::GetBatteryPercentageFromVoltage(uint16_t voltage) {
+  // The number of line segments used to approximate the battery discharge curve.
+  static const uint8_t LINE_SEGMENT_COUNT = 7;
+
+  // The voltages (mV) at the endpoints of the line segments. Any two consecutive
+  // values represent the start and end voltage of a line segment.
+  static const uint16_t voltageOffsets[LINE_SEGMENT_COUNT + 1] {4157, 4063, 3882, 3747, 3716, 3678, 3583, 3500};
+
+  // The battery percentages at the endpoints of the line segments. Note that last
+  // value is omitted: It is not needed because we only need the percentages at
+  // the start of each line segment.
+  static const float percentageOffsets[LINE_SEGMENT_COUNT] {100.000, 95.197, 70.429, 48.947, 35.158, 18.971, 5.801};
+
+  // The pre-calculated slopes (in battery percentage points per millivolt) of the
+  // line segments.
+  static const float percentageSlopes[LINE_SEGMENT_COUNT] {0.05109, 0.13684, 0.15913, 0.44481, 0.42595, 0.13863, 0.06989};
+
+  if (voltage >= voltageOffsets[0]) {
+    return 100;
+  }
+
+  if (voltage <= voltageOffsets[7]) {
```

Comment:
The 7 in this line seems hardcodes and maybe should be derived from LINE_SEGMENT_COUNT too, no?